GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift

```diff
@@ -0,0 +1,965 @@
+                let request = QLThumbnailGenerator.Request(
+                    fileAt: file.url,
+                    size: kSize,
+                    scale: UIScreen.main.scale,
+                    representationTypes: .lowQualityThumbnail
+                )
+                QLThumbnailGenerator.shared.generateRepresentations(for: request) { representation, _, _ in
+                    let image = representation?.uiImage
+                        ?? UIImage(contentsOfFile: file.url.path)?.resized(to: self.kSize)
+                    DispatchQueue.main.async {
+                        completion(image)
+                    }
+                }
+            } else {
+                let image = UIImage(contentsOfFile: file.url.path)?.resized(to: kSize)
+                completion(image)
+            }
+        }
+
+        enum ThumbnailState: Equatable {
+            case notGenerated
+            case inProgress(Kind.LocalFile)
+            case generated(image: UIImage, file: Kind.LocalFile)
+            case failed(Kind.LocalFile)
+        }
+
+        var imageState: ThumbnailState = .notGenerated {
+            didSet {
+                switch imageState {
+                case .notGenerated:
+                    break
+                case let .inProgress(file):
+                    previewImage(for: file) { [weak self] image in
+                        switch image {
+                        case let .some(img):
+                            self?.imageState = .generated(image: img, file: file)
+                        case .none:
+                            self?.imageState = .failed(file)
+                        }
+                    }
+                case let .failed(file):
+                    self.setPreviewImage(nil, for: file)
+                case let .generated(image, file):
+                    self.setPreviewImage(image, for: file)
+                }
+            }
+        }
+    }
+}
+
+extension SecureConversations.FileUploadListView.FileUploadView.FilePreviewView.ThumbnailState {
+    var isLoaded: Bool {
+        switch self {
+        case .generated:
+            return true
+        default:
+            return false
+        }
+    }
+}
+
+extension SecureConversations.FileUploadListView.FileUploadView.FilePreviewView.Kind {
+    struct LocalFile: Equatable {
+        var fileExtension: String
+        var fileName: String
+        let fileSize: Int64?
+        let fileSizeString: String?
+        let fileInfoString: String?
+        let isImage: Bool
+        let url: URL
+    }
+}
+
+private extension FilePreviewStyle {
+    static let initial = FileUploadStyle.initial.filePreview
+}
+
+private extension FileUploadStyle {
+    static let initial = Theme().chat.messageEntry.uploadList.item
+}
+
+// TODO: Add unit tests
+// MOB-1815
+/// Collection that allows easier access by element identifier.
+struct IdCollection<Identifier: Hashable, Item: Equatable>: Equatable & Collection {
```

Comment:
Would it be reasonable to move it to separate file?

Reply:
Very reasonable 41274fac4e0a17be2e089747b4086c92b1c2934a.